pkg/controllers/dnspolicy/dnspolicy_healthchecks.go

```diff
@@ -2,203 +2,181 @@ package dnspolicy
 
 import (
 	"context"
-	"crypto/md5"
 	"fmt"
-	"io"
+	"time"
 
-	"k8s.io/apimachinery/pkg/api/equality"
+	"github.com/kuadrant/kuadrant-operator/pkg/common"
+	"github.com/kuadrant/kuadrant-operator/pkg/reconcilers"
+	"github.com/miekg/dns"
+	"github.com/openshift/library-go/pkg/controller"
+	"k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
 	crlog "sigs.k8s.io/controller-runtime/pkg/log"
-	gatewayv1beta1 "sigs.k8s.io/gateway-api/apis/v1beta1"
-
-	"github.com/kuadrant/kuadrant-operator/pkg/reconcilers"
 
+	"github.com/Kuadrant/multicluster-gateway-controller/pkg/_internal/slice"
 	"github.com/Kuadrant/multicluster-gateway-controller/pkg/apis/v1alpha1"
-	"github.com/Kuadrant/multicluster-gateway-controller/pkg/dns"
-	"github.com/Kuadrant/multicluster-gateway-controller/pkg/traffic"
 )
 
-// healthChecksConfig represents the user configuration for the health checks
-type healthChecksConfig struct {
-	Endpoint         string
-	Port             *int64
-	FailureThreshold *int64
-	Protocol         *dns.HealthCheckProtocol
-}
-
 func (r *DNSPolicyReconciler) reconcileHealthChecks(ctx context.Context, dnsPolicy *v1alpha1.DNSPolicy, gwDiffObj *reconcilers.GatewayDiff) error {
-	log := crlog.FromContext(ctx)
-
-	// Delete Health checks for each gateway no longer referred by this policy
-	for _, gw := range gwDiffObj.GatewaysWithInvalidPolicyRef {
-		log.V(1).Info("reconcileHealthChecks: gateway with invalid policy ref", "key", gw.Key())
-		_, err := r.reconcileGatewayHealthChecks(ctx, gw.Gateway, nil)
+	var expectedProbes []*v1alpha1.DNSHealthCheckProbe
+	if dnsPolicy.DeletionTimestamp != nil {
+		expectedProbes = []*v1alpha1.DNSHealthCheckProbe{}
+	} else {
+		var err error
+		expectedProbes, err = r.generateExpectedProbes(ctx, gwDiffObj, dnsPolicy)
```

Comment:
I'd have expected this to keep the pattern of using the three groups of gateways from `gwDiffObj`  and calling an appropriate `reconcileGatewayHealthChecks` method to deal with create/updates/deletes of the probes. This is a pattern that was setup in this polices reconcilers (dnsrecord and health) based on other kuadrant polices see [RateLimit](https://github.com/Kuadrant/kuadrant-operator/blob/main/controllers/ratelimitpolicy_limits.go#L49-L87l)  and [AuthConfig](https://github.com/Kuadrant/kuadrant-operator/blob/main/controllers/authpolicy_istio_authorization_policy.go#L46-L53). I think we'd want good reason form straying from that pattern here since we are trying to keep all these polices as consistent as possible in how they work.

Reply:
I'm struggling to see a way to use that pattern, and have the resulting process be as reliable. The current code will look at all gateways referenced by the policy (new or existing) and generate the CRs as expected for those gateways based on the current DNS Policy configuration.

It will then take those and create or exist as expected, and remove any created as a result of this DNS Policy that are no longer expected.

As I have understood your comment, you're suggesting I change this to:
- Creates in a for loop for "newly targeted gateways"
- Updates in a for loop for "already targeted gateways"
- Deletes in a for loop for "no longer targeted gateways"

However, this feels inferior to me, as it leaves a greater possibility of the create / delete events to not be accurately triggered and have either new health checks not created, or expired health checks not removed. Have I misunderstood somewhere?

Reply:
oh, I think I see what you mean. Have it call a function per gateway which will generate a list of expected health checks per gateway and create / update / delete as expected, but treating each gateway separately?

Reply:
discussed on a call, I will update this to match the existing policy reconcilers.